Fixed #24914 -- Added basic permission mixins to contrib.auth

raphaelm opened the pull request:
https://code.djangoproject.com/ticket/24914

An initial version of implementation and tests is in the pull request, documentation is yet to be done. Comments on the code are very welcome.

This proposal provides all the and only the functionality of all three authentication decorators from contrib.auth.

While it is slightly differently implemented, it tries to stay compatible with the widely-used 3rd-party package django-braces in terms of parameter and method names. django-braces does have more functionality than the decorators have, but implementing this functionality in django itself is outside the scope of this ticket.

ewjoachim commented:
Saw a few things and told @raphaelm directly, looks all good to me now (except missing documentation ^^)

raphaelm commented:
I added documentation now. As I'm not a native speaker, I'd love someone to review the language.

I decieded to drop a section in the documentation on class-based view on wrapping as_view with decorators, as there no longer seems to be any decorator in django core where this would make any sense.

timgraham commented:
buildbot, add to whitelist.

raphaelm commented:
Fixed another small documentation problem right now.

raphaelm commented:
Fixed commit message

MarkusH commented:
@raphaelm How much of the django-braces stuff did you "copy" or how much is the code / tests inspired by them? Does it make sense to have a acknowledgement notice at the top of the files?

raphaelm commented:
@MarkusH I did not copy that much, as django-braces has an abstract ``AccessView`` and then derives the three mixins from that, whereas my patch uses UserPassesTestMixin as the base view and the other ones derive from that. I also changed some of smaller implementation details to make it more consistend with the decorators or to be just more readable. The tests are only inspired by the decorator tests from the django codebase.

However, acknowledging them seems nice. Is there a standard way to do this?

MarkusH commented on django/contrib/auth/mixins.py:
```diff
@@ -0,0 +1,91 @@
+        if not user_test_result:
+            return self.handle_no_permission(request)
+
+        return super(UserPassesTestMixin, self).dispatch(
```
You can join this and a couple of other lines. We have a max line length of 119 chars.

MarkusH commented on django/contrib/auth/mixins.py:
```diff
@@ -0,0 +1,91 @@
+        # In case the 403 handler should be called raise the exception
+        if self.raise_exception:
+            raise PermissionDenied
+        return redirect_to_login(request.get_full_path(),
```
Can you use hanging indents and a trailing comma, please:

```foo(
    1,
    2,
)
```

raphaelm commented:
@MarkusH I updated the pull request, both of your style comments should be fixed now. Thank you for pointing it out.

MarkusH commented on django/contrib/auth/mixins.py:
```diff
@@ -0,0 +1,92 @@
+    permission_required = None
+
+    def test_func(self, user):
+        if isinstance(self.permission_required, six.string_types):
```
Can you switch to the same logic as in the decorator: `if isinstance(..., (list, tuple)): ...`

raphaelm replied:
I can, absolutely, if you think I should. However, I'd rather adjust the decorator's logic to mine: has_perm will take any iterable, therefore restricting it to list and tuple does not seem logical to me. Passing a set can make perfect sense. Therefore, I used the same logic as Django uses in other places where you can pass an iterable or a string (like model._meta.ordering). What do you think?

Am 5. Juni 2015 00:34:33 GMT+01:00, schrieb Markus Holtermann <notifications@github.com>:
>> +
>> +    def test_func(self, user):
>> +        return user.is_authenticated()
>> +
>> +
>> +class PermissionRequiredMixin(UserPassesTestMixin):
>> +    """
>> +    Mixin for class-based views that checks whether a user has a
>particular permission
>> +    enabled, redirecting to the log-in page if necessary.
>> +    If the raise_exception parameter is given the PermissionDenied
>exception
>> +    is raised.
>> +    """
>> +    permission_required = None
>> +
>> +    def test_func(self, user):
>> +        if isinstance(self.permission_required, six.string_types):
>
>Can you switch to the same logic as in the decorator: `if
>isinstance(..., (list, tuple)): ...`
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/django/django/pull/4749/files#r31776720

MarkusH replied:
Valid point. Feel free to change the decorator in a separate commit.

MarkusH commented:
Regarding the acknowledgement, I'd add something like

```
The authentication mixins UserPassesTestMixin, LoginRequiredMixin and
PermissionRequiredMixin have been inspired by django-braces
<https://github.com/brack3t/django-braces/>
```

raphaelm commented:
I added the acknowledgement and I'm preparing the logic change for the decorator now, if you'd like to triage that: https://code.djangoproject.com/ticket/24929#ticket

MarkusH commented on django/contrib/auth/mixins.py:
```diff
@@ -0,0 +1,96 @@
+        if self.redirect_field_name is None:
+            raise ImproperlyConfigured(
+                '{0} has redirect_field_name set to None. Set {0}.redirect_field_name '
+                'to a strng or or override {0}.get_redirect_field_name().'.format(
```
"string"

MarkusH commented on docs/releases/1.9.txt:
```diff
@@ -70,6 +70,11 @@ Minor features
 
+* The module ``django.contrib.auth.mixins`` was added with ``LoginRequiredMixin`,
+  ``PermissionRequiredMixin`` and ``UserPassesTestMixin`` mixins to provide the
+  functionality of the `django.contrib.auth.decorators` shortcuts to class-based
```
double back-tics, please.

MarkusH commented on docs/releases/1.9.txt:
```diff
@@ -70,6 +70,11 @@ Minor features
   a deprecation warning in older versions and is no longer supported in
   Django 1.9).
 
+* The module ``django.contrib.auth.mixins`` was added with ``LoginRequiredMixin`,
```
Please link to the mixins:
```
:class:`~django.contrib.auth.mixins.LoginRequiredMixin`
```

MarkusH commented on docs/topics/auth/default.txt:
```diff
@@ -500,6 +500,40 @@ The login_required decorator
+The LoginRequired mixin
+~~~~~~~~~~~~~~~~~~~~~~~
+
+If you are using :doc:`class-based views </topics/class-based-views/index>`, you can
```
"s/If/When" I think.

MarkusH commented on docs/topics/auth/default.txt:
```diff
@@ -500,6 +500,40 @@ The login_required decorator
+        from django.contrib.auth.mixins import LoginRequiredMixin
+
+        class MyView(LoginRequiredMixin, View):
+            login_url = '/accounts/login/'
```
Can you change that to something different than settings.LOGIN_URL, e.g. `/login/`

MarkusH commented on docs/topics/auth/default.txt:
```diff
@@ -500,6 +500,40 @@ The login_required decorator
+    will be redirected to the login page.
+
+    If you set the ``login_url`` property on your view, this URL will be used
+    for the redirect. Otherwise, :setting:`settings.LOGIN_URL <LOGIN_URL>`
```
use
```
:settings:`~settings.LOGIN_URL`
```
to only display the variable name without the module.

MarkusH commented on docs/topics/auth/default.txt:
```diff
@@ -560,6 +594,32 @@ redirects to the login page::
+    When using :doc:`class-based views </topics/class-based-views/index>`, you can
+    use the ``UserPassesTestMixin`` to do this.
+
+    You should override the ``test_func`` method of the class to provide
```
You have to override ...

MarkusH commented on docs/topics/auth/default.txt:
```diff
@@ -500,6 +500,40 @@ The login_required decorator
+    will be used. By default, this URL will be called with a query string parameter
+    ``"next"`` containing the URL the user should be redirected to after a successful
+    login. You can change this name by setting the ``redirect_field_name`` parameter
+    on your view. Example::
```
I'd strip the "Example" and instead use "on your view::"

MarkusH commented on docs/topics/auth/default.txt:
```diff
@@ -560,6 +594,32 @@ redirects to the login page::
+    use the ``UserPassesTestMixin`` to do this.
+
+    You should override the ``test_func`` method of the class to provide
+    the actual test that is performed. Example::
```
Chop Example, as above.

MarkusH commented on docs/topics/auth/default.txt:
```diff
@@ -605,10 +665,30 @@ Applying permissions to generic views
+    This mixin, just like the ``permisison_required``
+    decorator, checks whether the user accessing a view has a particular
+    permission. You should specify the permission (or a list or permissions)
+    using the ``permission_required`` parameter. Example::
```
Chop Example, as above

MarkusH commented:
There's a bunch of docs duplication with regards to `login_url`, `raise_exception` and `redirect_field_name`. I'd document them on the `UserPassesTestMixin` and reference their documentation from `LoginRequiredMixin` and `PermissionRequiredMixin` instead.

MarkusH commented on django/contrib/auth/mixins.py:
```diff
@@ -0,0 +1,96 @@
+    the class method `test_func`.
+    """
+    login_url = None
+    redirect_field_name = REDIRECT_FIELD_NAME
```
Can you sort those attributes please

MarkusH commented on django/contrib/auth/mixins.py:
```diff
@@ -0,0 +1,96 @@
+    def get_test_func(self):
+        return getattr(self, "test_func")
+
+    def dispatch(self, request, *args, **kwargs):
```
I'd keep `dispatch()` the first method on the view.

MarkusH commented on tests/auth_tests/test_mixins.py:
```diff
@@ -0,0 +1,125 @@
+@override_settings(ROOT_URLCONF='auth_tests.urls')
+class LoginRequiredMixinTestCase(AuthViewsTestCase):
+    """
+    Tests the LoginRequired mixin
```
I don't think this docstring is required.

MarkusH commented on tests/auth_tests/test_mixins.py:
```diff
@@ -0,0 +1,125 @@
+    """
+    Tests the LoginRequired mixin
+    """
+    def setUp(self):
```
There's a class method called `setUpTestData()`. Please use that instead.

MarkusH commented on tests/auth_tests/test_mixins.py:
```diff
@@ -0,0 +1,125 @@
+        self.user = models.User.objects.create(username='joe', password='qwerty')
+        self.factory = RequestFactory()
+
+    def testLoginRequired(self, view=None, login_url=None):
```
use snake case, please: `test_login_required()`

MarkusH commented:
I'm missing tests for the `UserPassesTestMixin`.

raphaelm commented:
I updated the pull request and fixed/changed everything you remarked, except those two:

> There's a bunch of docs duplication with regards to login_url, raise_exception and redirect_field_name. I'd document them on the UserPassesTestMixin and reference their documentation from LoginRequiredMixin and PermissionRequiredMixin instead.

I am unsure whether this is the best solution. I feel that UserPassesTestMixin is the least important one of those and most users will only be interested in LoginRequiredMixin and maybe PermissionRequiredMixin. Therefore, I think it would be more confusing then helpful.

> I'm missing tests for the UserPassesTestMixin.

Me too, but how would you test this properly? UserPassesTest only makes sense to use if you override test_func and test that subclass -- however, this is exactly what the tests for LoginRequiredMixin and PermissionRequiredMixin do. All aspects of UserPassesTest are tested, but I do not see a nice way to do it directly.

(There is no test for the user_passes_test decorator, I guess for the same reason. Also, most mistakes you pointed out in my code in your last comments, are present in the decorators as well.)

raphaelm commented:
Fixed spelling now, but the docs build still fails because of broken links. I do not see the mistake, do you see it?

raphaelm commented:
Okay, I found the problem, documentation builds now.

timgraham commented:
FWIW, I'd give acknowledgement in AUTHORS and/or the commit message. That seems more in line with the 4 bullet here: https://docs.djangoproject.com/en/1.8/internals/contributing/writing-code/coding-style/#miscellaneous

MarkusH commented:
Regarding documentation duplication: The arguments are implemented on the `UserPassesTestMixin`, which I think implies that they are documented there.

MarkusH commented on tests/auth_tests/test_mixins.py:
```diff
@@ -0,0 +1,121 @@
+        request.user = AnonymousUser()
+        response = view(request)
+        self.assertEqual(response.status_code, 302)
+        self.assertIn(login_url, response.url)
```
Isn't there something like `assertRedirects()`?

raphaelm replied:
Yes, but apparently it only works with the test client, not with RequestFactory. I'll change the assertIn to an equality check, but I don't think it is worth creating a bunch of URLs for the test views in the URL config just to test these mixins.

MarkusH commented:
Regarding the missing tests: I'd create 2 mixins in the tests that have `def test_func(self, user): return True` (`False` respectively) and check all the combinations of `login_url`, `raise_exception`, and `redirect_field_name`. The `LoginRequiredMixin` and `PermissionRequiredMixin` could then just check one positive and one negative use case.

raphaelm commented:
Okay, I'll look into this during the next days.

Am 6. Juni 2015 16:12:32 MESZ, schrieb Markus Holtermann <notifications@github.com>:
>Regarding the missing tests: I'd create 2 mixins in the tests that have
>`def test_func(self, user): return True` (`False` respectively) and
>check all the combinations of `login_url`, `raise_exception`, and
>`redirect_field_name`. The `LoginRequiredMixin` and
>`PermissionRequiredMixin` could then just check one positive and one
>negative use case.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/django/django/pull/4749#issuecomment-109582290

raphaelm commented:
> FWIW, I'd give acknowledgement in AUTHORS and/or the commit message.

I'd go for the commit message than as its only a few lines coming from there.

I updated the documentation and tests according to all your other comments.

raphaelm commented:
Just updated the PR again, now rebased to current ``master`` without any merge conflicts.

MarkusH commented on tests/auth_tests/test_mixins.py:
```diff
@@ -0,0 +1,147 @@
+
+
+class LoginRequiredMixinTestCase(AuthViewsTestCase):
+    @classmethod
```
One empty line above, please.

MarkusH commented on docs/topics/auth/default.txt:
```diff
@@ -560,6 +595,47 @@ redirects to the login page::
+
+    .. versionadded:: 1.9
+
+    When using :doc:`class-based views </topics/class-based-views/index>`, you can
```
Can you wrap the documentation at 79 chars, please.

MarkusH commented on tests/auth_tests/test_mixins.py:
```diff
@@ -0,0 +1,147 @@
+
+
+class PermissionsRequiredMixinTest(TestCase):
+    @classmethod
```
As above.

MarkusH commented on django/contrib/auth/mixins.py:
```diff
@@ -0,0 +1,85 @@
+
+class PermissionRequiredMixin(UserPassesTestMixin):
+    """
+    Mixin for class-based views that checks whether a user has a particular permission
```
Wrap at 79 chars, please.

raphaelm commented:
Style issues fixed, PR updated.

MarkusH commented:
Replaced by #4836
